Core/include/Acts/Seeding/CandidatesForSpM.hpp

```diff
@@ -9,141 +9,136 @@
-
-  template<typename external_space_point_t>
-  inline bool CandidatesForSpM<external_space_point_t>::exists(std::size_t n, std::size_t max_size) const
-  { return n < max_size; }
-
-  template<typename external_space_point_t>
-  inline float CandidatesForSpM<external_space_point_t>::weight(const std::vector<value_type>& storage, std::size_t n) const
-  { return std::get<InternalComponents::I_WEIGHT>(storage[n]); }
-
-  template<typename external_space_point_t>
-  inline void CandidatesForSpM<external_space_point_t>::clear()
-  {
-    // do not clear max size, this is set only once
-    m_n_high = 0;
-    m_n_low = 0;
-    // clean fixed space points
-    m_SpB = default_value;
-    m_SpM = default_value;
-    // clean storage
-    m_storage_high.clear();
-    m_storage_low.clear();
-  }
-  
+template <typename external_space_point_t>
+class CandidatesForSpM {
+  // Internal collection, we store only the components that are different and
+  // useful
+  enum InternalComponents : int { I_BSP = 0, I_TSP, I_WEIGHT, I_ZORIGIN };
+
+ public:
+  // more complete collection of variables, used by external seeding code
+  enum Components : int { BSP = 0, MSP, TSP, WEIGHT, ZORIGIN, QUALITY };
+
+  using sp_type = external_space_point_t*;
+  using value_type = std::tuple<sp_type, sp_type, float, float>;
+  using output_type = std::tuple<sp_type, sp_type, sp_type, float, float, bool>;
+  static constexpr sp_type default_value = nullptr;
+
+  CandidatesForSpM();
+  ~CandidatesForSpM() = default;
+
+  void setMaxElements(std::size_t n_low, std::size_t n_high);
+  void setMediumSp(sp_type);
+  void setBottomSp(sp_type);
+  std::vector<output_type> storage() const;
+  const sp_type& spM() const;
+
+  void push(sp_type& SpT, float weight, float zOrigin, bool isQuality);
+  void clear();
+
+ private:
+  bool exists(std::size_t, std::size_t) const;
+
+  void pop(std::vector<value_type>&, std::size_t&);
+  float top(const std::vector<value_type>&) const;
+  float weight(const std::vector<value_type>&, std::size_t) const;
+
+  void bubbleup(std::vector<value_type>&, std::size_t);
+  void bubbledw(std::vector<value_type>&, std::size_t, std::size_t);
+
+  void addToCollection(std::vector<value_type>&, sp_type& SpB, sp_type& SpT,
+                       float weight, float zOrigin, bool isQuality);
+  void insertToCollection(std::vector<value_type>&, sp_type& SpB, sp_type& SpT,
+                          float weight, float zOrigin, bool isQuality);
+
+ public:
+  // sizes
+  std::size_t m_max_size_high;
+  std::size_t m_max_size_low;
+  std::size_t m_n_high;
+  std::size_t m_n_low;
+
+  // space points
+  sp_type m_SpB;
+  sp_type m_SpM;
+
+  // storage
+  // These vectors are sorted as a min heap tree
+  // Each node is lower then its childs
+  // Thus, it is guaranteed that the lower elements is at the front
+  // Sorting criteria is the seed quality
+
+  // storage for candidates with high quality
+  std::vector<value_type> m_storage_high;
```

Comment:
I wonder if the std::set with ordering by "quality" would not be optimal here.

Reply:
Hi @tboldagh, that is something I can try as well and compare with my current implementation. Indeed I was hoping to receive feedbacks exactly on using more efficient data structures. Thanks!

However I believe that `std::set` are implemented as balanced trees with the minimum value being the further left node. Insertion and deletion are log(N) operations (as in my implementation) as well as lookup for the minimum (while in my current implementation minimum lookup is a constant operation since it's always the first element in the vector)

Reply:
These log(N) are realy relevant at large N. Mostly irrelevant for small collections. Maybe still be worth trying.

Reply:
A quick test of mine with an ad-hoc code suggests that `std::set` may be slower. I believe this is due to the fact that in a `std::set` elements are often allocated dynamically as they are added and there is no requirement for items to be contiguous in memory, as for a `std::vector`

Reply:
OK If the point is empirically verified then it is settled. Maybe in the doc of this class it would be worth saying something along these lines:
"This is in effect faster sorted container - implementation with std::set was tried and found slower." 
so that people do not waste time in future as I managed you to do! Apologies Carlo!